[UnifiedPDF][iOS] Create plugin specific viewport configuration.

sammygill opened the pull request:
#### 4aebedeb66526aa469f882e0fae0e91b50833f99
<pre>
[UnifiedPDF][iOS] Create plugin specific viewport configuration.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281167">https://bugs.webkit.org/show_bug.cgi?id=281167</a>
<a href="https://rdar.apple.com/137627221">rdar://137627221</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

Main frame PDFs iOS currently have their viewport configuration set to ViewportConfiguration::webpageParameters
which uses defaultDesktopViewportWidth as the width of the content&apos;s viewport. This ends
up causing a number of issues that restrict us from getting any sort of scales correct
as we end up incorrectly sizing the plugin. In this patch we create a viewport configuration
that is specific to plugins so that we can tailor the parameters as appropriate for this
specific case.

This configuration is mostly the same as nativeWebpageParameters with the maximum scale
set to 1. We do this to avoid accidentally setting the page scale factor to any other
scales since the plugin is completely responsible for managing its scale. We may need to
fine tune this configuration a bit more in the future but this should hopefully serve as
a sane initial set that we can build upon.

We set this configuration when we initialize the plugin in PluginView::setParent by
informing the web page that we have initialized the plugin. The web page will then set
the viewport configuration to this new plugin specific one.

* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::ViewportConfiguration::pluginParameters):
* Source/WebCore/page/ViewportConfiguration.h:
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setParent):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::resetViewportDefaultConfiguration):
(WebKit::WebPage::didInitializePlugin):

Canonical link: <a href="https://commits.webkit.org/284976@main">https://commits.webkit.org/284976@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf05eea7665af12cc53a3bbe30ba47ee1cd36b5b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71016 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23789 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75120 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22221 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73132 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58227 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22039 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56162 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14632 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74082 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45809 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61216 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36602 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18651 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20562 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64398 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19011 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76840 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15248 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18201 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63891 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15292 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61276 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63851 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15733 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11950 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5594 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46229 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1005 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47301 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48584 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47043 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/ViewportConfiguration.cpp:
```diff
@@ -415,6 +415,21 @@ ViewportConfiguration::Parameters ViewportConfiguration::nativeWebpageParameters
+    parameters.allowsUserScaling = true;
+    parameters.allowsShrinkToFit = false;
+    parameters.minimumScale = 1;
+    parameters.maximumScale = 1;
```
Hm.. `allowsUserScaling=true` alongside `min/maximumScale=1` seems strange. I suppose this is because we want the user to be able to pinch zoom the scroll view, but then intercept any subsequent scaling and handle it purely within the plugin?

sammygill replied:
Yeah this was carried over form the native web page configuration, but after taking a look at how its used I agree that it's a bit strange and we *probably* don't want it with the behavior we intend to implement for the plugin

aprotyas commented on Source/WebCore/page/ViewportConfiguration.h:
```diff
@@ -145,6 +145,7 @@ class ViewportConfiguration {
     WEBCORE_EXPORT Parameters nativeWebpageParameters();
     static Parameters nativeWebpageParametersWithoutShrinkToFit();
     static Parameters nativeWebpageParametersWithShrinkToFit();
+    WEBCORE_EXPORT static Parameters pluginParameters();
```
I was expecting the name to read more like `pluginDocumentParameters`

sammygill replied:
I think that makes sense to me as well!

sammygill committed:
[UnifiedPDF][iOS] Create plugin specific viewport configuration.
https://bugs.webkit.org/show_bug.cgi?id=281167
rdar://137627221

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

Main frame PDFs iOS currently have their viewport configuration set to ViewportConfiguration::webpageParameters
which uses defaultDesktopViewportWidth as the width of the content's viewport. This ends
up causing a number of issues that restrict us from getting any sort of scales correct
as we end up incorrectly sizing the plugin. In this patch we create a viewport configuration
that is specific to plugins so that we can tailor the parameters as appropriate for this
specific case.

This configuration is mostly the same as nativeWebpageParameters with the maximum scale
set to 1. We do this to avoid accidentally setting the page scale factor to any other
scales since the plugin is completely responsible for managing its scale. We may need to
fine tune this configuration a bit more in the future but this should hopefully serve as
a sane initial set that we can build upon.

We set this configuration when we initialize the plugin in PluginView::setParent by
informing the web page that we have initialized the plugin. The web page will then set
the viewport configuration to this new plugin specific one.

* Source/WebCore/page/ViewportConfiguration.cpp:
(WebCore::ViewportConfiguration::pluginParameters):
* Source/WebCore/page/ViewportConfiguration.h:
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::setParent):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::resetViewportDefaultConfiguration):
(WebKit::WebPage::didInitializePlugin):

Canonical link: https://commits.webkit.org/284976@main
